src/graph/Edge.java

```diff
@@ -38,12 +35,31 @@ public void reverse() {
   }
 
   public Edge reversed() {
-    Edge e = this.copy();
+    Edge e = (Edge) this.clone();
     e.reverse();
     return e;
   }
 
   @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || this.getClass() != o.getClass()) {
```

Comment:
Normally, people use (! o instanceof Edge).

Reply:
What should I do if `o` is a subclass of `Edge`?

Reply:
We do not have this use case now.